File: services/risk-ml-service/app/core/health.py
"""Health check module for risk-ml-service."""
from typing import Any, Callable, Optional
import asyncio
from enum import Enum
from dataclasses import dataclass, field

from fastapi import FastAPI, Response
from pydantic import BaseModel


class HealthStatus(str, Enum):
    UP = "up"
    DOWN = "down"


class CheckResult(BaseModel):
    name: str
    status: HealthStatus
    error: Optional[str] = None


class HealthResponse(BaseModel):
    status: HealthStatus
    checks: dict[str, CheckResult] = {}
    details: dict[str, Any] = {}


@dataclass
class HealthCheck:
    name: str
    check: Callable[[], Any]
    timeout: float = 5.0


@dataclass
class HealthChecker:
    checks: list[HealthCheck] = field(default_factory=list)
    details: dict[str, Any] = field(default_factory=dict)
    _ready: bool = False

    def add_check(
        self, name: str, check: Callable[[], Any], timeout: float = 5.0
    ) -> None:
        self.checks.append(HealthCheck(name=name, check=check, timeout=timeout))

    def set_detail(self, key: str, value: Any) -> None:
        self.details[key] = value

    def set_ready(self, ready: bool) -> None:
        self._ready = ready

    def is_ready(self) -> bool:
        return self._ready

    async def _run_check(self, check: HealthCheck) -> CheckResult:
        try:
            result = check.check()
            if asyncio.iscoroutine(result):
                await asyncio.wait_for(result, timeout=check.timeout)
            return CheckResult(name=check.name, status=HealthStatus.UP)
        except asyncio.TimeoutError:
            return CheckResult(
                name=check.name,
                status=HealthStatus.DOWN,
                error=f"Timeout after {check.timeout}s",
            )
        except Exception as e:
            return CheckResult(
                name=check.name, status=HealthStatus.DOWN, error=str(e)
            )

    async def run_checks(self) -> HealthResponse:
        results = await asyncio.gather(
            *[self._run_check(check) for check in self.checks]
        )

        checks_dict = {r.name: r for r in results}
        overall_status = HealthStatus.UP

        for result in results:
            if result.status == HealthStatus.DOWN:
                overall_status = HealthStatus.DOWN
                break

        return HealthResponse(
            status=overall_status, checks=checks_dict, details=self.details
        )


def register_health_routes(app: FastAPI, checker: HealthChecker) -> None:
    @app.get("/health")
    async def health():
        response = await checker.run_checks()
        status_code = 200 if response.status == HealthStatus.UP else 503
        return Response(
            content=response.model_dump_json(),
            status_code=status_code,
            media_type="application/json",
        )

    @app.get("/health/live")
    async def liveness():
        return {"status": "alive"}

    @app.get("/health/ready")
    async def readiness():
        if not checker.is_ready():
            return Response(
                content='{"status": "not_ready"}',
                status_code=503,
                media_type="application/json",
            )

        response = await checker.run_checks()
        status_code = 200 if response.status == HealthStatus.UP else 503
        return Response(
            content=response.model_dump_json(),
            status_code=status_code,
            media_type="application/json",
        )

File: services/alert-service/pkg/health/health.go
package health

import (
	"context"
	"net/http"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
)

// Status represents health check status
type Status string

const (
	StatusUp   Status = "up"
	StatusDown Status = "down"
)

// CheckFunc is a function that performs a health check
type CheckFunc func(ctx context.Context) error

// Check represents a single health check
type Check struct {
	Name    string
	Check   CheckFunc
	Timeout time.Duration
}

// CheckResult represents the result of a health check
type CheckResult struct {
	Name   string `json:"name"`
	Status Status `json:"status"`
	Error  string `json:"error,omitempty"`
}

// Response represents the health check response
type Response struct {
	Status  Status                 `json:"status"`
	Checks  map[string]CheckResult `json:"checks,omitempty"`
	Details map[string]any         `json:"details,omitempty"`
}

// Checker manages health checks
type Checker struct {
	mu       sync.RWMutex
	checks   []Check
	details  map[string]any
	ready    bool
	readyMu  sync.RWMutex
}

// NewChecker creates a new health checker
func NewChecker() *Checker {
	return &Checker{
		checks:  make([]Check, 0),
		details: make(map[string]any),
		ready:   false,
	}
}

// AddCheck adds a health check
func (c *Checker) AddCheck(name string, check CheckFunc, timeout time.Duration) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.checks = append(c.checks, Check{
		Name:    name,
		Check:   check,
		Timeout: timeout,
	})
}

// SetDetail sets a detail value
func (c *Checker) SetDetail(key string, value any) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.details[key] = value
}

// SetReady sets the readiness state
func (c *Checker) SetReady(ready bool) {
	c.readyMu.Lock()
	defer c.readyMu.Unlock()
	c.ready = ready
}

// IsReady returns the readiness state
func (c *Checker) IsReady() bool {
	c.readyMu.RLock()
	defer c.readyMu.RUnlock()
	return c.ready
}

// RunChecks executes all health checks
func (c *Checker) RunChecks(ctx context.Context) Response {
	c.mu.RLock()
	checks := make([]Check, len(c.checks))
	copy(checks, c.checks)
	details := make(map[string]any)
	for k, v := range c.details {
		details[k] = v
	}
	c.mu.RUnlock()

	results := make(map[string]CheckResult)
	overallStatus := StatusUp

	var wg sync.WaitGroup
	resultsCh := make(chan CheckResult, len(checks))

	for _, check := range checks {
		wg.Add(1)
		go func(ch Check) {
			defer wg.Done()

			timeout := ch.Timeout
			if timeout == 0 {
				timeout = 5 * time.Second
			}

			checkCtx, cancel := context.WithTimeout(ctx, timeout)
			defer cancel()

			result := CheckResult{
				Name:   ch.Name,
				Status: StatusUp,
			}

			if err := ch.Check(checkCtx); err != nil {
				result.Status = StatusDown
				result.Error = err.Error()
			}

			resultsCh <- result
		}(check)
	}

	wg.Wait()
	close(resultsCh)

	for result := range resultsCh {
		results[result.Name] = result
		if result.Status == StatusDown {
			overallStatus = StatusDown
		}
	}

	return Response{
		Status:  overallStatus,
		Checks:  results,
		Details: details,
	}
}

// LivenessHandler returns the liveness probe handler
func (c *Checker) LivenessHandler() gin.HandlerFunc {
	return func(ctx *gin.Context) {
		ctx.JSON(http.StatusOK, gin.H{
			"status": "alive",
		})
	}
}

// ReadinessHandler returns the readiness probe handler
func (c *Checker) ReadinessHandler() gin.HandlerFunc {
	return func(ctx *gin.Context) {
		if !c.IsReady() {
			ctx.JSON(http.StatusServiceUnavailable, gin.H{
				"status": "not_ready",
			})
			return
		}

		response := c.RunChecks(ctx.Request.Context())

		status := http.StatusOK
		if response.Status == StatusDown {
			status = http.StatusServiceUnavailable
		}

		ctx.JSON(status, response)
	}
}

// HealthHandler returns the combined health handler
func (c *Checker) HealthHandler() gin.HandlerFunc {
	return func(ctx *gin.Context) {
		response := c.RunChecks(ctx.Request.Context())

		status := http.StatusOK
		if response.Status == StatusDown {
			status = http.StatusServiceUnavailable
		}

		ctx.JSON(status, response)
	}
}

// RegisterRoutes registers health check routes
func (c *Checker) RegisterRoutes(router *gin.Engine) {
	router.GET("/health", c.HealthHandler())
	router.GET("/health/live", c.LivenessHandler())
	router.GET("/health/ready", c.ReadinessHandler())
}
